src/Microsoft.Data.SqlClient/netcore/src/Common/src/System/Net/Security/NegotiateStreamPal.Unix.cs

```diff
@@ -156,20 +157,29 @@ internal static partial class NegotiateStreamPal
             }
 
             SafeFreeNegoCredentials negoCredentialsHandle = (SafeFreeNegoCredentials)credentialsHandle;
+            SecurityStatusPal status = default;
 
-            if (negoCredentialsHandle.IsDefault && string.IsNullOrEmpty(spn))
+            foreach (string spn in spns)
             {
-                throw new PlatformNotSupportedException(Strings.net_nego_not_supported_empty_target_with_defaultcreds);
-            }
+                if (negoCredentialsHandle.IsDefault && string.IsNullOrEmpty(spn))
+                {
+                    throw new PlatformNotSupportedException(Strings.net_nego_not_supported_empty_target_with_defaultcreds);
+                }
 
-            SecurityStatusPal status = EstablishSecurityContext(
-                negoCredentialsHandle,
-                ref securityContext,
-                spn,
-                requestedContextFlags,
-                ((inSecurityBufferArray != null && inSecurityBufferArray.Length != 0) ? inSecurityBufferArray[0] : null),
-                outSecurityBuffer,
-                ref contextFlags);
+                status = EstablishSecurityContext(
+                    negoCredentialsHandle,
+                    ref securityContext,
+                    spn,
+                    requestedContextFlags,
+                    ((inSecurityBufferArray != null && inSecurityBufferArray.Length != 0) ? inSecurityBufferArray[0] : null),
+                    outSecurityBuffer,
+                    ref contextFlags);
+
+                if (status.ErrorCode != SecurityStatusPalErrorCode.InternalError)
+                    break;
+                else 
+                    securityContext = null; // Reset security context to be generated again
```

Comment:
This securityContext is passed as reference from caller. Can we save its initial status and assign it back if we need to try another SPN?

Reply:
It's reset to null when trying another SPN, so it can be generated again for next SPN. Old one failed with error, hence not needed to be saved.